scrape/scrape.go

```diff
@@ -838,8 +925,16 @@ loop:
 			}
 			break
 		}
-		if et != textparse.EntrySeries {
+		switch et {
+		case textparse.EntryType:
+			sl.cache.setType(p.Type())
+			continue
+		case textparse.EntryHelp:
+			sl.cache.setHelp(p.Help())
```

Comment:
Where is the Help getting unescaped? I didn't see it in the lexing code.

Reply:
This is problematic since we'd allocate a new string. If we store it unescaped, we'd need to unescape all help strings on each scrape and do the same. If we store escaped and unescaped version, our memory usage for this feature doubles, since it's virtually all in help strings.

We could only do it on demand when reading the help strings via the API – that will also come with heavy allocs, but is certainly better.
Only for everything to get escaped all over again when we return it as JSON... 

Any ideas on how to handle this all?

There's also no generic unescape function AFAICS. So we have to do replacements ourselves. Is there a full list of escaped symbols we've to cover?

Reply:
https://golang.org/pkg/strconv/#Unquote

We've already got code for this for label values, might be worth seeing how it works. The vast majority of strings won't need escaping.

For help strings it's only \\ and \n.

Reply:
Yes, I realized it shortly after I posted. The docs don't quite give it away, the docs for the symmetric `Quote` are clearer. It expects actual quotes at beginning and end though, which means we need to allocate yet another transient string to add those.

Reply:
Ah right, we did the whole thing for label values already. The notable difference is that it's called only on first sight a series there. We'd be doing it on every scrape – so if there's escaped stuff in there, the impact is notably bigger.

Reply:
Tabs can be represented by themselves, so no need to escape them.

Reply:
It seems rather unusual. I don't think I've encountered anything in the wild yet the opted to only escape some things but not others. And I'm relatively sure it's gonna throw of more common escape handling (Go's JSON encoder errors for example) – after all this parser is not the only consumer.

Reply:
If that's the case then those parsers are buggy. \t for example should end up as a backslash followed by a t.